test/integration/tcp_proxy_integration_test.cc

```diff
@@ -1293,14 +1298,25 @@ TEST_P(TcpProxyDynamicMetadataMatchIntegrationTest, DynamicMetadataMatch) {
   initialize();
 
   expectEndpointToMatchRoute([](IntegrationTcpClient& tcp_client) -> std::string {
-    // Break the write into two; validate that the first is received before sending the second. This
-    // validates that a downstream filter can use this functionality, even if it can't make a
-    // decision after the first `onData()`.
-    EXPECT_TRUE(tcp_client.write("p", false));
-    tcp_client.waitForData("p");
+    // Break the write into multiple chunks; validate that the first word is received before sending
+    // the rest. This validates that a downstream filter can use this functionality, even if it
+    // can't make a decision after the first `onData()`.
+    EXPECT_TRUE(tcp_client.write("pri", false));
+    // Writing additional data in multiple chunks to show that buffering of early data in tcp_buffer
+    // works properly
+    EXPECT_TRUE(tcp_client.write("mary is ", false));
```

Comment:
This is not the early data buffer you create in Tcp proxy. It is a little bit confusing for the comment, like we stop the filter chain iteration at `InjectDynamicMetadata::onData()` when just receives `pri`, which Tcp proxy cannot get the `onData` call before the `InjectDynamicMetadata` continue the filter chain.

And writing a new set of integration tests maybe more clear here to make sure the original behavior is also tested. We should also add some situations that trigger `early_data_buffer_` when connection is not created, but seems not easy to test.

Reply:
@jrajahalme please address this.